src/dfx/src/commands/deps/pull.rs

```diff
@@ -0,0 +1,340 @@
+    canisters_to_pull: &mut VecDeque<Principal>,
+    pulled_json: &mut PulledJson,
+) -> DfxResult {
+    info!(
+        logger,
+        "Resolving dependencies of canister {canister_id}..."
+    );
+
+    match fetch_metatdata(agent, canister_id, DFX_DEPS).await {
+        Ok(Some(deps_raw)) => {
+            let deps_str = String::from_utf8(deps_raw)?;
+            let mut deps = vec![];
+            for entry in deps_str.split_terminator(';') {
+                match entry.split_once(':') {
+                    Some((_, p)) => {
+                        let dep_id = Principal::from_text(p)
+                            .with_context(|| format!("`{p}` is not a valid Principal."))?;
+                        canisters_to_pull.push_back(dep_id);
+                        deps.push(dep_id);
+                    }
+                    None => bail!(
+                        "Failed to parse `dfx:deps` entry: {entry}. Expected `name:Principal`. "
+                    ),
+                }
+            }
+            pulled_json.canisters.insert(
+                canister_id,
+                PulledCanister {
+                    deps,
+                    ..Default::default()
+                },
+            );
+        }
+        Ok(None) => {
+            warn!(
+                logger,
+                "`{DFX_DEPS}` metadata not found in canister {canister_id}."
+            );
+            pulled_json
+                .canisters
+                .insert(canister_id, PulledCanister::default());
+        }
+        Err(e) => {
+            bail!(e);
+        }
+    }
+    Ok(())
+}
+
+// download canister wasm, candid, init
+async fn download_canister_files(
+    env: &dyn Environment,
+    logger: &Logger,
+    canister_id: Principal,
+    pulled_canister: &mut PulledCanister,
+) -> DfxResult {
+    info!(logger, "Pulling canister {canister_id}...");
+
+    let agent = env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+
+    // try fetch `dfx:wasm_hash`. If not available, get the hash of the on chain canister.
+    let hash_on_chain = match fetch_metatdata(agent, canister_id, DFX_WASM_HASH).await {
+        Ok(Some(wasm_hash_raw)) => {
+            let wasm_hash_str = String::from_utf8(wasm_hash_raw)?;
+            trace!(
+                logger,
+                "Canister {canister_id} specified a custom hash: {wasm_hash_str}"
+            );
+            hex::decode(wasm_hash_str)?
+        }
+        Ok(None) => {
+            let canister_status =
+                get_canister_status(env, canister_id, &CallSender::SelectedId).await?;
+            match canister_status.module_hash {
+                Some(hash_on_chain) => hash_on_chain,
+                None => {
+                    bail!("Canister {canister_id} doesn't have module hash. Perhaps it's not installed.");
+                }
+            }
+        }
+        Err(e) => {
+            bail!(e);
+        }
```

Comment:
Almost anytime Err(e) => bail!(e) means might as well check `?` outside of the match
```suggestion
```